### DISCUSSION TITLE
Register an api description without handler

### DISCUSSION THREAD

**Discussion by clintjedwards** [danielgtaylor/huma#366] [Q&A] [answered]
Hello!

I'm not sure if this is possible, but I figured I would ask. I'm attempting to transition a GRPC API into using REST. So far Huma is making that really easy to get good documentation and a spec out of the migration. I'm loving it so far, thank you.

One feature of GRPC I've run into that is causing me issues is the ability to have a bi-directional stream with a client. To solve this I've picked up websockets, but (to my knowledge) all websockets tooling expects access to the underlying `http.ResponseWriter` and `http.Request` objects such that it may manage the request writes and reads.

I searched a bit, but it does not seem like Huma gives this access to these objects, so I figured no big deal, I'll just add the handler manually to my router. But if I do that then I don't get the great autogenerated openapi spec from huma for that specific route.

I also tried adding an empty handler func via huma.Register to get the documentation and then added (at the end) a `router.HandleFunc` with the same path I gave the register command to see if I could overwrite the handler. This didn't work and weirdly enough no matter which position the default router handler was in put in, the huma handler was run instead.

I was wondering if there is a way to do one of two things:

1. Register a route without a handler so that I may register the handler through the native router API.
2. Overwrite the current handler for a particular route (I'll be honest that I was too lazy to look deeper into this part, but I assume that huma just adds on to the default router's route table)

Or anything else that might help on this front. I'd also be willing to help with a PR if you have a feature in mind you just don't have the time for.

Thanks!

**Comment by bclements** [comment]
Wouldn’t you want to do this using SSE instead of webhooks? I may be
misunderstanding your use case.

**Reply by clintjedwards** [reply]
My understanding is that SSE is unidirectional, I'm setting up webhooks because ideally I'd like the server to push some events, then the client, then server, and so on.

**Comment by bclements** [comment]
Correct then you’d be looking at websockets not webhooks.

**Comment by bclements** [comment]
I am idiot, I assume this is for service to service communication so
webhooks is appropriate. I will go back to lurking now. :)

**Comment by clintjedwards** [comment]
No @bclements, actually you're completely correct. I am the idiot. I typed this after searching for a solution for several hours so I assume I was just tired, but I actually meant websockets not webhooks. I'm not sure why my brain typed webhooks several times.

I've edited the original post to actually say websockets now. Sorry for the confusion.

Thanks for the gut check.

**Comment by bclements** [comment]
Webhooks can also be used for service to service communication but limited
use cases. So you technically weren't wrong.

**Comment by danielgtaylor** [comment, marked as answer]
@clintjedwards you have full access to the OpenAPI so can modify it as needed. Here's a quick runnable example: https://go.dev/play/p/GGkYMi3Cn26

```go
api.OpenAPI().Paths["/demo"] = &huma.PathItem{
	Get: &huma.Operation{
		OperationID: "get-demo",
		Summary:     "Get a demo",
		Responses: map[string]*huma.Response{
			"200": {
				Description: "OK",
				Content: map[string]*huma.MediaType{
					"application/json": {
						Schema: &huma.Schema{Type: huma.TypeString},
					},
				},
			},
		},
	},
}
```

It's a little extra work, but you can manually describe any operation registered with the underlying router that way. You can also use the schema registry at `api.OpenAPI().Components.Schemas` to get schemas for your types to put into the OpenAPI, including generating `$ref`s as needed. Hope this helps!

**Comment by clintjedwards** [comment]
Oh man it seems so obvious now in retrospect. Many thanks, sorry for the simple question.

**Comment by davidolrik** [comment]
@danielgtaylor The way you are adding them here assumes that you have previously added another endpoint, if you haven't you'll get a `panic: assignment to entry in nil map`

The correct or [preferred](https://github.com/danielgtaylor/huma/blob/77c7a16c4c84444fa626305c14216fb9a02a887f/openapi.go#L1460-L1464) way to do it would be via `AddOperation()`  like so:

```go
api.OpenAPI().AddOperation(
	&huma.Operation{
		OperationID: "get-demo",
		Method: http.MethodGet,
		Path: "/demo",
		Summary: "Get a demo",
		Responses: map[string]*huma.Response{
			"200": {
				Description: "OK",
				Content: map[string]*huma.MediaType{
					"application/json": {
						Schema: &huma.Schema{Type: huma.TypeString},
					},
				},
			},
		},
	},
)
```

I got bitten by random panics as I had previously done as you suggested above.

Because the load order of my handlers are non deterministic, sometimes the one with the direct access to `api.OpenAPI().Paths["..."]` would get called first resulting in a panic.
